workflow-api/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    /// automatically contain all signals defined with the `#[signal]` attribute.
+    fn signals() -> &'static [&'static SignalDefinition<Self>] {
+        // TODO
+        &[]
+    }
+    /// All queries this workflow can handle. Typically you won't implement this directly, it will
+    /// automatically contain all queries defined with the `#[query]` attribute.
+    fn queries() -> &'static [&'static QueryDefinition<Self>] {
+        // TODO
+        &[]
+    }
+}
+
+/// TODO: Exists in SDK in slightly different form, and would move into this crate
+#[derive(Debug)]
+pub enum WfExitValue<T: TemporalSerializable> {
+    /// Continue the workflow as a new execution
+    ContinueAsNew(Box<ContinueAsNewWorkflowExecution>), // Wouldn't be raw proto in reality
+    /// Confirm the workflow was cancelled
+    Cancelled,
+    /// Finish with a result
+    Normal(T),
+}
+impl<T: TemporalSerializable> From<T> for WfExitValue<T> {
+    fn from(v: T) -> Self {
+        Self::Normal(v)
+    }
+}
+// ... also convenience functions for constructing C-A-N, etc.
+
+/// A workflow context which contains only information, but does not allow any commands to
+/// be created.
+pub struct SafeWfContext {
+    // TODO
+}
+
+/// TODO: Placeholder, exists in SDK and would move into this crate & (likely) become a trait
+pub struct WfContext {}
+impl WfContext {
+    pub async fn timer(&self, _: Duration) {
+        todo!()
+    }
+}
+
+pub struct SignalDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The handler input type must be erased here, since otherwise we couldn't store/return the
+    // heterogeneous collection of definition types in the workflow itself. The signal macro
+    // will wrap the user's function with code that performs deserialization.
+    handler: Box<dyn FnMut(&mut WF, Payload) -> Result<(), anyhow::Error>>,
+}
+pub struct QueryDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The query macro will wrap the user's function with code that performs deserialization of
+    // input and serialization of output, as well as error boxing.
+    handler: Box<dyn FnMut(&WF, Payload) -> Result<Payload, anyhow::Error>>,
+}
+
+/// TODO: Placeholders, likely belong inside protos crate. These will be auto-implemented for
+///   anything using serde already (which I expect is how virtually everyone will do this).
+pub trait TemporalSerializable {}
+impl<T> TemporalSerializable for T {}
+pub trait TemporalDeserializable {}
+impl<T> TemporalDeserializable for T {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::FutureExt;
+    use std::{collections::HashMap, marker::PhantomData};
+
+    // Workflow implementation example
+    struct MyWorkflow {
+        foo: u64,
+        bar: HashMap<String, u64>,
+    }
+
+    impl Workflow for MyWorkflow {
```

Comment:
```suggestion
    #[workflow]
    impl Workflow {
```

I wonder if there's value in not requiring trait impl and instead marking the run method with `#[workflow_run]` to extract input/output types. Only reason I was thinking this is because I kinda like that I can put my signal/query/update functions in the same `impl` block as my run stuff. But otherwise, I see how it hurts `new()` (unless you put the attribute there too).

Reply:
People are pretty used to multiple impl blocks in Rust. That said, I can get behind a procmacro like this too if it's providing value like auto-naming. It just needs to not do too much magic, because messing up IDEs etc is where things get really bad.